### PR TITLE
Always use UTF8 as the encoding for Kotlin code, regardless of the system/default charset

### PR DESCRIPTION
**Pull request by nreid260**
Closes https://github.com/facebook/ktfmt/issues/373

### CODE REVIEW

**Comment by nreid260** [comment]
@cgrushko I'm not sure I've set up the Maven stuff right. In order to test the default charset is ignored, I want to run the tests using UTF16, which requires passing a JVM startup flag.

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt [hunk 1]
```diff
@@ -40,7 +41,7 @@ data class ParsedArgs(
 
     fun processArgs(err: PrintStream, args: Array<String>): ParsedArgs {
       if (args.size == 1 && args[0].startsWith("@")) {
-        return parseOptions(err, File(args[0].substring(1)).readLines().toTypedArray())
+        return parseOptions(err, File(args[0].substring(1)).readLines(UTF_8).toTypedArray())
```
Comment (cgrushko) [inline core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:44]: according to https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/java.io.-file/read-lines.html , UTF_8 is the default, no?

File: core/src/main/java/com/facebook/ktfmt/cli/Main.kt [hunk 2]
```diff
@@ -20,18 +20,22 @@ import com.facebook.ktfmt.format.Formatter
...
 import java.util.concurrent.atomic.AtomicInteger
 import kotlin.system.exitProcess
 
 class Main(
-    private val input: InputStream,
-    private val out: PrintStream,
-    private val err: PrintStream,
+    private val stdin: InputStream,
+    private val stdout: PrintStream,
+    private val stderr: PrintStream,
```
Comment (cgrushko) [inline core/src/main/java/com/facebook/ktfmt/cli/Main.kt:38]: can we leave this to separate PR? lots of noise in the diff...

**Review by cgrushko — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt (hunk 1 above)
Comment (nreid260) [reply core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:44]: I think it pays to be explicit. All the Java APIs (some of which are still available in Kotlin) are ambiguous about what their default will be.

File: core/src/main/java/com/facebook/ktfmt/cli/Main.kt (hunk 2 above)
Comment (nreid260) [reply core/src/main/java/com/facebook/ktfmt/cli/Main.kt]: Done

File: core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt (hunk 1 above)
Comment (cgrushko) [reply core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:44]: but - doesn't that mean we already always use utf8?

Comment (nreid260) [reply core/src/main/java/com/facebook/ktfmt/cli/ParsedArgs.kt:44]: When parsing the args file, yes. That was mostly an afterthought of fixing the issues with Kotlin source files.

**Comment by cgrushko** [comment]
I feel this PR is doing multiple things; let's split into multiple PRs.
1. Fix the exact issue at hand - please don't change stderr into stdout or vice verse, some tools rely on text going into one or the other.
2. Add explicit UTF_8 everywhere - my vote is to avoid that. Kotlin uses utf8 by default. A test would be enough to catch problems. (same as we wouldn't explicitly use utf8 in golang).

**Comment by nreid260** [comment]
1. I've reverted the stdout writes, but it's weird. stdout now emits in two different charsets depending on the type of output.
2. Being explicit is the main point of this PR. The encoding was implicit when the code was originally written, so humans didn't think about what the behaviour should be, which lead to bugs. I want to trigger that consideration next time.

File: core/src/main/java/com/facebook/ktfmt/cli/Main.kt [hunk 3]
```diff
@@ -158,18 +163,14 @@ class Main(
       err.println("Error formatting $fileName: ${e.message}; skipping.")
       throw e
     } catch (e: ParseError) {
-      handleParseError(fileName, e)
+      err.println("$fileName:${e.message}")
       throw e
     } catch (e: FormattingError) {
       for (diagnostic in e.diagnostics()) {
-        System.err.println("$fileName:$diagnostic")
+        err.println("$fileName:$diagnostic")
```
Comment (hick209) [inline core/src/main/java/com/facebook/ktfmt/cli/Main.kt:170]: Thank you!

**Review by hick209 — APPROVED** [review]

**Comment by hick209** [comment]
Maybe... 😅

I guess me @cgrushko, @davidtorosyan, @strulovich and @cortinico are the non-official-but-maybe-official-ish maintainers of ktfmt and it depends on who has bandwidth to tackle things.

I was trying to keep track and clean things up here, but then I went on paternity leave in September. Now that I'm back and was able to catch up, I'm trying to put some love into the project, at the very least managing GitHub issues and PRs.

You could also try to use the [Slack channel](https://kotlinlang.slack.com/archives/C01GZCU0QNB) we have tor conversations in case you don't want to do everything through GitHub as well, I guess only @davidtorosyan is not there.

**Comment by hick209** [comment]
Locally I see a bunch of failing tests from the class `com.facebook.ktfmt.cli.MainTest`.
@nreid260, would you mind rebasing and taking a look if you repro it as well?

**Comment by nreid260** [comment]
> Locally I see a bunch of failing tests from the class `com.facebook.ktfmt.cli.MainTest`. @nreid260, would you mind rebasing and taking a look if you repro it as well?

It looks like only the plugin tests are failing, and that's been happening since f64da02ad7a683fe10a3fc33fbc338c831dfc0a9

**Comment by hick209** [comment]
Are you saying that it passes on your local tests?
Could it maybe be related to the fact that we use buck to test it and not Maven, which means we are not using the plugin you set in https://github.com/facebook/ktfmt/blob/b394fe3fadb50aa8110bb291d1a46a9ea5a3b367/core/pom.xml#L140-L148 ?

This is what I get from running it locally from this PR and I've just rebased and checked that the base was clear
```
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - --dry-run prints nothing when there are no changes needed (file) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - all files in args are processed, even if one of them has an error (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - Using '-' as the filename formats an InputStream (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - --dry-run prints 'stdin' and does not reformat code from stdin (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - Exit code is 0 when there are changes (file) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - dropbox-style is passed to formatter (file) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - formatting from stdin prints formatted code to stdout regardless of whether it was already formatted (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - --set-exit-if-changed and --dry-run changes nothing, prints filenames, and exits with 1 (file) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - Exits with 0 when there are changes (stdin) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - Exit code is 1 when there are changes and --set-exit-if-changed is set (file) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - --dry-run prints nothing when there are no changes needed (stdin) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - Parsing errors are reported (file) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - Exit code is 1 when there are changes and --set-exit-if-changed is set (stdin) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - expandArgsToFileNames - single arg which is not a file is not returned (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - dropbox-style is passed to formatter (stdin) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - file is modified if it is not formatted (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - Always use UTF8 encoding (file) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - expandArgsToFileNames - multiple directory args are resolved to their recursively contained kt files (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - expandArgsToFileNames - a dash is an error (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - file is not modified if it is already formatted (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - Parsing errors are reported (stdin-name) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - --set-exit-if-changed and --dry-run changes nothing, prints filenames, and exits with 1 (stdin) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - expandArgsToFileNames - resolves 'kt' and 'kts' filenames only (recursively) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - --dry-run prints filename and does not change file (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - --stdin-name can only be used with stdin (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - Parsing errors are reported (stdin) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - Always use UTF8 encoding (stdin, stdout) (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - expandArgsToFileNames - single file arg is used as is (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
✗ Fail: fbsource//xplat/ktfmt/core/src/test/java/com/facebook/ktfmt/cli:cli - expandArgsToFileNames - single arg which is a directory is resolved to its recursively contained kt files (com.facebook.ktfmt.cli.MainTest) (0.1s)
expected: UTF-16
but was : UTF-8
at com.facebook.ktfmt.cli.MainTest.setUp(MainTest.kt:51)
```

**Comment by hick209** [comment]
Would it be possible to rewrite the tests in a way that does not rely on that Maven plugin?

**Comment by nreid260** [comment]
In Bazel I wrote the test with `jvm_flags = ["-Dfile.encoding=UTF-16"],`. Maven is not my primary build system.

**Comment by hick209** [comment]
Yep, that did it! No action item needed from you them, I now just need the sign in from someone from inside to merge this in

**Pull request closed without merging**